src/main/io/osd.c

```diff
@@ -225,7 +225,7 @@ static uint8_t osdGetHeadingIntoDiscreteDirections(int heading, int directions)
 {
     // Split input heading 0..359 into sectors 0..(directions-1), but offset
     // by half a sector so that sector 0 gets centered around heading 0.
-    heading = (heading * 2 + 360 / directions) % 720;
+    heading = ((heading * 2 + 360 / directions) + 720) % 720;
```

Comment:
This has become quite convoluted and difficult to follow now. Why not rewrite it to do it step by step:

```
#define FULL_CIRCLE 360
[...]
const unsigned segment = FULL_CIRCLE / directions;
heading = (heading + FULL_CIRCLE) % FULL_CIRCLE; // normalise
heading = heading + (segment / 2) % FULL_CIRCLE; // shift by half a segment
return heading / segment; // return segment number
```

Reply:
Yes, I will do it ASAP.

Reply:
@mikeller one thing: the original code, used FULL_CIRCLE*2 in some operations to maintain precision when making divisions (to not loose the ,5 of the division, increasing precision).

For example, the `FULL_CIRCLE / directions = 360 / 16 = 22,5` 

The number of directions "must" be always  pair, so multiplying by two all the operations will solve this problem.

Reply:
If precision is a concern here, then `heading` should be a float. Multiplying by some magic number that relies on an implied assumption is not a proper solution.

Reply:
The "solution" with the "magic number" will be like this:
```
    heading = (heading + FULL_CIRCLE) % FULL_CIRCLE; // normalise

    // Split input heading 0..359 into sectors 0..(directions-1), but offset
    // by half a sector so that sector 0 gets centered around heading 0.

    // We use a FULL_CIRCLE * 2 to not loose precision in divisions
    const unsigned segment = (FULL_CIRCLE * 2) / directions;
    heading = ((heading * 2) + (segment / 2)) % (FULL_CIRCLE * 2); // shift by half a segment

    return (heading / segment); // return segment number
```
In this way, we work with a 720 circle and not a 360.

The problem is specially the latest division, `heading / segment` can be quite different if segment is 22 or 22,5, and can return different segment number.  But using a float I suppose is less efficient that multiplying by two, I don't know. 

I have the two solutions ready, I can upload any of them. I don't know how important is the precision here, maybe your original solution will be enough.

Reply:
I think a solution will be to multiply by the `directions` and not by the magic number 2. This will maintain the maximum precision using integer operations. Tomorrow I will try to study this option.

Reply:
@McGiverGim : Yes, simply substitute `segment = FULL_CIRCLE / directions` into following equations and than simplify them. Unless I made some mistake, it should be:
```
int segment = (heading * directions + 180) / 360; 
return segment % directions;        // normalize result, assuming heading >=0. 
```
if `heading` can be < 0, add at least max(directions) * 360 (256*360 may be safe choice) in first equation before division.

Reply:
That's pretty much what I was going to propose last night, before I hit 'Start new review' instead of 'Comment' (but then, it was 2am after all):

```
#define FULL_CIRCLE 360
[...]
const unsigned segment = FULL_CIRCLE;
heading = ((heading * directions) + FULL_CIRCLE) % (FULL_CIRCLE * directions); // normalise
heading = (heading + (FULL_CIRCLE / 2)) % (FULL_CIRCLE * directions); // shift by half a segment
return heading / FULL_CIRCLE; // return segment number
```

Also, `directions` should be `unsigned`.

Reply:
@mikeller I think your normalisation does not need to multiply by directions, if you do that adding FULL_CIRCLE will be not enough to let the number positive. I've the code ready, similar to yours, I will upload it in some minutes.